doc/dev/crimson/wait_states.rst

```diff
@@ -0,0 +1,84 @@
+===========
+wait states
+===========
+
+In crimson, exactly like in the classical OSD, a client request has data and
+ordering dependencies which must be satisfied before processing (actually
+a particular phase of) can begin. An obvious example of such data dependency
+is the fact that an OSD needs to have a version of OSDMap that matches the one
+used by the client (``Message::get_min_epoch()``).
+
+If a dependency is not satisfied, the processing stops. It is crucial to note
+the same must happen to all other requests that are sequenced-after (due to
+their ordering requirements).
+
+There are a few cases when the blocking of a client request can happen.
+
+
+  ``ClientRequest::ConnectionPipeline::await_map``
+    wait for particular OSDMap version is available at the OSD level
+  ``ClientRequest::ConnectionPipeline::get_pg``
+    wait a particular PG becomes available on OSD
+  ``ClientRequest::PGPipeline::await_map``
+    wait on a PG being advanced to particular epoch
+  ``ClientRequest::PGPipeline::wait_for_active``
+    wait on a PG becomes ``is_active()``
+  ``ClientRequest::PGPipeline::recover_missing``
+    wait on an object has been recovered
+  ``ClientRequest::PGPipeline::get_obc``
+    wait on an object context becomes locked
+  ``ClientRequest::PGPipeline::process``
+    wait if any other ``MOSDOp`` message is handled against this PG
+
+At any moment an instance of ``ClientRequest`` can be only in one of these
+phases. Similarly, an object denoting particular phase can host not more
+than single ``ClientRequest`` the same time. At low-level this is achieved
+through lock a `seastar::shared_mutex` instance which is the heat of any
+of a such phase-denoting object.
+
+As the execution advances,
+request enters next phase and leaves the current one freeing it for another
+``ClientRequest`` instance. All these phases form a pipeline which assures
+the order is preserved.
```

Comment:
These phases are chained in the order of execution just like stages in a pipeline. This machinery fulfills the data dependencies required for serving a request.

@rzarzynski i think the order preserving is a more complicated topic. 

1. what is the order, at which scale, order by what?
2. why we need to order. what is the contract?
3. how to we preserve the ordering.

Reply:
> what is the order, at which scale, order by what?
> how to we preserve the ordering.

Extended the paragraph about the `ConnectionPipeline` and `PGPipeline`. Covered the flow where requests from multiple connections target same PG.

> why we need to order. what is the contract?

Added a sentence about keeping the classical OSD's semantic on the beginning of the doc.